clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2803,7 +2803,207 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
   return {Template, AliasRhsTemplateArgs};
 }
 
-// Build deduction guides for a type alias template.
+// Build deduction guides for a type alias template from the given underlying
+// deduction guide F.
+FunctionTemplateDecl *
+BuildDeductionGuideForTypeAlias(Sema &SemaRef,
```

Comment:
Best I can tell this is mostly a Copy/paste from the other two locations?  Can you point out for me what are the meaningful different changes?

Reply:
(if possible, please put the non-functional changes into their own commit for review purposes)

Reply:
We just move the implementation from `DeclareImplicitDeductionGuidesForTypeAlias ` to a dedicated function, so that we can reuse it in the `DeclareAggregateDeductionGuideForTypeAlias`.

The functional change of this commit is in `DeclareAggregateDeductionGuideForTypeAlias`, the previous implementation is incorrect, and we should use the one in `DeclareImplicitDeductionGuidesForTypeAlias`.

Reply:
I have restructured commits in this PR. It contains two commits: one for the NFC refactoring change, the other one is for the actual fix. I hope it would make the review easier.

Reply:
to clarify, in the actual fix you removed a whole bunch of code?
https://github.com/llvm/llvm-project/pull/90894/commits/621899224aa7d59ab70675f6ff32b298b06a515a